File: src/dataset/typescript/twelve-days_Exercism-typescript-FilipeCerejo/twelve-days.ts
type VerseType = {
    day: string;
    gift: string;
};

type SongType = {
    song: string;
    gifts: string[];
};

const TWELVEDAY: VerseType[] = [
    { day: 'first', gift: 'a Partridge in a Pear Tree.' },
    { day: 'second', gift: 'two Turtle Doves, and' },
    { day: 'third', gift: 'three French Hens,' },
    { day: 'fourth', gift: 'four Calling Birds,' },
    { day: 'fifth', gift: 'five Gold Rings,' },
    { day: 'sixth', gift: 'six Geese-a-Laying,' },
    { day: 'seventh', gift: 'seven Swans-a-Swimming,' },
    { day: 'eighth', gift: 'eight Maids-a-Milking,' },
    { day: 'ninth', gift: 'nine Ladies Dancing,' },
    { day: 'tenth', gift: 'ten Lords-a-Leaping,' },
    { day: 'eleventh', gift: 'eleven Pipers Piping,' },
    { day: 'twelfth', gift: 'twelve Drummers Drumming,' },
];

export function recite(begin: number, end: number): string {
    let song: SongType = { song: '', gifts: [] };

    TWELVEDAY.slice(0, end).forEach((verse: VerseType, idx: number) => {
        song.gifts.unshift(verse.gift);
        if (idx >= begin - 1) {
            song.song += `On the ${TWELVEDAY[idx].day} day of Christmas my true love gave to me: ${song.gifts.join(
                ' '
            )}\n`;
        }
    });
    return song.song;
}

File: src/dataset/Java/twelve-days_exercism-java-blogscot/TwelveDays.java
import java.util.stream.Collectors;
import java.util.stream.IntStream;

class TwelveDays {

  private static final String[] days = {
      "first",
      "second",
      "third",
      "fourth",
      "fifth",
      "sixth",
      "seventh",
      "eighth",
      "ninth",
      "tenth",
      "eleventh",
      "twelfth"};

  private static final String[] gifts = {
      "and a Partridge in a Pear Tree",
      "two Turtle Doves",
      "three French Hens",
      "four Calling Birds",
      "five Gold Rings",
      "six Geese-a-Laying",
      "seven Swans-a-Swimming",
      "eight Maids-a-Milking",
      "nine Ladies Dancing",
      "ten Lords-a-Leaping",
      "eleven Pipers Piping",
      "twelve Drummers Drumming"};

  private static String createVerse(String day, String gifts) {
    return String.format("On the %s day of Christmas my true love gave to me: %s.\n", day, gifts);
  }

  private static String getGifts(int numGifts) {
    if (numGifts == 0) return "a Partridge in a Pear Tree";

    return IntStream.rangeClosed(0, numGifts)
        .map(i -> numGifts - i) // count downTo
        .mapToObj(i -> gifts[i])
        .collect(Collectors.joining(", "));
  }

  String verse(int verseNumber) {
    var num = verseNumber - 1;
    return createVerse(days[num], getGifts(num));
  }

  String verses(int startVerse, int endVerse) {
    return IntStream.rangeClosed(startVerse, endVerse)
        .mapToObj(this::verse)
        .collect(Collectors.joining("\n"));
  }

  String sing() {
    return verses(1, 12);
  }
}
